src/steamship/agents/functional/functions_based.py

```diff
@@ -67,10 +72,8 @@ def build_chat_history_for_tool(self, context: AgentContext) -> List[Block]:
         # this should happen BEFORE any agent/assistant messages related to tool selection
         messages.append(context.chat_history.last_user_message)
```

Comment:
Am I right that this will result in the following message list:

- System Message
- Topic-Selected Ancient Messages In Order
- Last User Message
- Last Action

I feel like if I was the LLM, I would want to see at the very least the last K messages from the history, in order, and the topic-selected list shouldn't include any that were included in the Last K messages list.

Reply:
We have:

1. System message
2. Semantic match historical messages
3. Context-window messages (user <--> assistant and currently assistant<-->tool)
4. Last User message
5. Scratchpad messages.

(2) and (3) should be de-duped and sorted by file index (which is a current placeholder for temporal order).

So, I think that we are meeting your notional LLM requirements.